packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -86,6 +115,38 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
     return result;
   }
 
+  /// Stores the engine layer created for this layer in order to reuse engine
+  /// resources across frames for better app performance.
+  ///
+  /// Set this field to the value returned by one of [ui.SceneBuilder]'s "push"
+  /// methods, e.g. [ui.SceneBuilder.pushOpacity].
+  ///
+  /// This value may be passed to [ui.SceneBuilder.addRetained] to communicate
+  /// to the engine that nothing in this layer or any of its descendants
+  /// changed. The native engine could, for example, reuse the texture rendered
+  /// in a previous frame. The web engine could, for example, reuse the HTML
+  /// DOM nodes created for a previous frame.
+  ///
+  /// This value may be passed as `oldLayer` argument to a "push" method to
+  /// communicate to the engine that a layer is updating a previously rendered
+  /// layer. The web engine could, for example, update the properties of
+  /// previously rendered HTML DOM nodes rather than creating new nodes.
+  ui.EngineLayer get engineLayer => _engineLayer;
+  @protected
+  set engineLayer(ui.EngineLayer newLayer) {
+    _engineLayer = newLayer;
+    if (!alwaysNeedsAddToScene) {
+      // We created a new engine layer for this layer. Therefore this layer no
+      // longer requires that its [addToScene] is called (e.g. it is now safe
+      // to add this layer as retained or use it as `oldLayer`). However, the
+      // parent must construct a new engine layer to add this layer to, and so
+      // we mark it as needing its [addToScene] called next frame.
```

Comment:
what if this layer's child layer called markNeedsAddToScene on us for that exact reason?

Reply:
That is expected. This is designed to handle two situations:

1. When rendering the complete layer tree as normal. In this case we call child `addToScene` methods first, then we call `set engineLayer` for the parent. The children will call `markNeedsAddToScene` on the parent to signal that they produced new engine layers and therefore the parent needs to update. However, in this case, the parent is already updating, and so after it's done, _its_ `set engineLayer` is called and it clears the `_needsAddToScene` flag.

2. When rendering an interior layer (e.g. `OffsetLayer.toImage`). In this case we call `addToScene` for one of the children but not the parent, i.e. we produce new engine layers for children but not for the parent. Here the children will mark the parent as needing `addToScene`, but the parent does not clear the flag until some future frame decides to render it, at which point the parent knows that it cannot retain its engine layer and will call `addToScene` again.

Reply:
Please document all that in the source. The existing comment wasn't clear enough for me to get that from it.

It's sad that toImage results in us throwing away the "real" engine layers.